docs/architecture/adr-021-abci-events.md

```diff
@@ -0,0 +1,47 @@
+# ADR 012: ABCI Events
+
+## Changelog
+
+- *2018-09-02* Remove ABCI errors component. Update description for events
+- *2018-07-12* Initial version
+
+## Context
+
+ABCI tags were first described in [ADR 002](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-002-event-subscription.md).
+They are key-value pairs that can be used to index transactions.
+
+Currently, ABCI messages return a list of tags to describe an
+"event" that took place during the Check/DeliverTx/Begin/EndBlock,
+where each tag refers to a different property of the event, like the sending and receiving account addresses.
+
+Since there is only one list of tags, recording data for multiple such events in
+a single Check/DeliverTx/Begin/EndBlock must be done using prefixes in the key
+space.
+
+TODO: brief description of how the indexing works
+
+## Decision
+
+Instead of returning a list of tags, return a list of events, where
+each event is a list of tags. This way we naturally capture the concept of
+multiple events happening during a single ABCI message.
+
+TODO: describe impact on indexing and querying
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Ability to track distinct events separate from ABCI calls (DeliverTx/BeginBlock/EndBlock)
+- More powerful query abilities
+
+### Negative
+
+- More complex query syntax
```

Comment:
I don't think the query syntax needs to change. We don't need to allow searching for transactions by combinations of *multiple* events - "slashed AND jailed" - they can just be treated independently, like multiple separate transactions would be at the moment.

Reply:
Maybe I'm misunderstanding, the `logical and` would say only return events that fit both slashing and the jailed tags. We still want that right? What we don't need is to support `logical or` imo. (E.g. all queries that are staking or are governance, as these can be independent) we want the former case to mitigate bandwidth for lite clients

Reply:
refs https://github.com/tendermint/tendermint/issues/1421